### PR TITLE
Add Carthage support

### PR DESCRIPTION
**Pull request by victorgama**
This adds carthage support. Some structural changes were needed, as I stated on #182, but none of them actually changes the source code (basically renamed an `xcodeproj`, and added another)

### CODE REVIEW

**Comment by fpillet** [comment]
Thanks for submitting this, Victor! Let me review this and I'll merge.

**Comment by victorgama** [comment]
I would recommend you wait until Carthage 0.12 is released, if you don't mind.

**Comment by fpillet** [comment]
Sure, what's the reason? I've not followed Carthage development - I mostly use CocoaPods, but I'm happy to have support merged in!

**Comment by victorgama** [comment]
Looks like it is trying to compile the wrong `xcodeproj`, and I couldn't figure out why.
Or even worse, it might not be Carthage's entire fault: `xcodebuild` is hanging while processing `Desktop Viewer/NSLogger.xcodeproj`. We could solve this by moving the Desktop Viewer out of the library repository, but I don't think it is a good idea, you see: One of Carthage's developer said it should select the root-most `xcodeproj` file, but for some really weird reason, it ignores the _root-most_, and ends up selecting the other file I already mentioned.

Anyway I'm still trying to sort this out without messing with your repository structure. :)

**Comment by victorgama** [comment]
@fpillet Okay, finally figured out the problem behind `xcodebuild` hanging. After I renamed one of the `xcodeproj` files, thinking it had something to do with Carthage picking based on a name (I was wrong about that), the whole `xcuserdatad` directory was ignored by `.gitignore`. Adding them back solved the issue and I could successfully build it through Carthage. I see you are already making releases through tags, so we're set. I think this can be merged now.

Please notice Carthage also supports distribution through prebuilt binaries. You can check it out [here](https://github.com/Carthage/Carthage#supporting-carthage-for-your-framework), but it work without it as well.

Thanks!

**Comment by oflannabhra** [comment]
:+1: @victorgama!
Thanks for the work! I was just getting ready to do this myself when I realized a PR already existed. I am very interested in being able to use this.

**Comment by victorgama** [comment]
@oflannabhra Carthage FTW 💖:shipit:

**Comment by victorgama** [comment]
@oflannabhra would you mind to try using `victorgama/nslogger` as a source in your Cartfile? I would like to know if you run into any issue. :grinning:

**Comment by oflannabhra** [comment]
trying right now! I think I might be running into issues. Carthage doesn't produce a lot of output, but I'm not sure it is finding the right scheme to build for NSLogger.

**Comment by victorgama** [comment]
@oflannabhra Hmm. Just tried here and it built without problems. Mind sharing your output? You can checkout mine [here](https://gist.github.com/victorgama/1aba62699c0d703ef11e).

**Comment by oflannabhra** [comment]
Could you copy your Cartfile? I've tried the following:
```
github "victorgama/NSLogger"
github "victorgama/NSLogger" "test"
github "victorgama/NSLogger" "carthage-support"
github "victorgama/NSLogger "d147e9d9456f6428067ff9a901ae27e5ff44cd39"
```
and they all hang on checking out. 

One really weird thing is that if I run
`carthage update` carthage will create an empty `xcodebuild` log file, but if I use 
`carthage update --verbose` carthage hangs at `*** Checking out NSLogger at xxx`

**Comment by victorgama** [comment]
@oflannabhra It's something related to shared schemes. `xcodebuild` just hangs when listing them!
You can check for yourself by running `xcodebuild -project <project path> -list`. `ps aux | grep xcodebuild` will also show you the whole command.

Another problem: `ctrl+c`ing Carthage will leave `xcodebuild` running until you manually kills it or restarts your machine.

**Comment by oflannabhra** [comment]
Yep, looks like Carthage is grabbing the `NSLogger/Desktop Viewer/NSLoggerViewer.xcodeproj` instead of the `NSLogger/NSLogger.xcodeproj` that you created. I verified that the `NSLogger.xcodeproj` that you created does exist in the checked out repository.

Looks like this is an issue with Carthage. I'm using `v0.11.0`.

**Comment by oflannabhra** [comment]
hah! I see you have already chimed in on [1047](https://github.com/Carthage/Carthage/issues/1047), @victorgama 

How did you resolve that? You've got it building properly, right?

**Comment by victorgama** [comment]
Yeah, I was trying to figure out why would Carthage pick `Desktop Viewer` over the _root-most_ file, mentioned on 1047. Ended up with nothing at all.

Right now I'm trying to move the shared schema to the `.xcworkspace` file, but now the result is:

```
*** Skipped building nslogger due to the error:
Dependency "nslogger" has no shared framework schemes
```

Also, they are expecting to release `0.12` in the next week or so.

Another option is @fpillet move the Desktop Viewer to its own repository (along with other tools), keeping only files related to the library itself here, but again, I don't think it is worth it. What you think, @oflannabhra?

**Comment by victorgama** [comment]
@oflannabhra http://stackoverflow.com/a/20941812 also mentions the fix I used.

**Comment by oflannabhra** [comment]
I'm still having the issue with Carthage grabbing the wrong xcodeproj. I probably will not pursue it until Carthage v 0.12 ships.

I don't think NSLogger should change its repository structure just for Carthage support, so until Carthage fixes its issues I think we will have to hold off on merging this PR.

**Comment by victorgama** [comment]
I tried to build it from source, but GitHub went down and I gave up. Will try to build from master later today.

**Comment by fpillet** [comment]
I agree with @oflannabhra, if this is going to break NSLogger builds I won't merge it. This is too bad because I'd love to have Carthage support - but will come soon anyway.

**Comment by victorgama** [comment]
@fpillet Next week :sparkles:

**Comment by victorgama** [comment]
Ping @oflannabhra! 0.12 was released 2 days ago. Mind trying to compile with my fork? It should automatically fetch the `v1.6.7` tag. :)

**Comment by oflannabhra** [comment]
:cry: 
```
*** Fetching NSLogger
*** Checking out NSLogger at "v1.6.7"
Failed to discover shared schemes in project NSLoggerViewer.xcodeproj — either the project does not have any shared schemes, or xcodebuild timed out.

If you believe this to be a project configuration error, please file an issue with the maintainers at https://github.com/victorgama/NSLogger/issues/new
```

**Comment by oflannabhra** [comment]
@victorgama looks like carthage `0.12` is still grabbing the wrong project.

**Comment by victorgama** [comment]
@oflannabhra Considering carthage/carthage#1047, what do we do? Wait until next release?

**Comment by oflannabhra** [comment]
I'm not sure... I'm waiting on this PR to make a working PR that adds Carthage Support to [XCDLumberjackNSLogger](https://github.com/0xced/XCDLumberjackNSLogger). I'd always prefer to do that... Is there a way to modify `NSLoggerViewer.xcodeproj` so that `xcodebuild` fails instead of timing out? Would that even let Carthage continue to the next build (`NSLogger.xcodeproj`)?

**Comment by victorgama** [comment]
@oflannabhra Not sure. As I said, Carthage actually builds it fine for me (I'm already using NSLogger through cartage by using the fork as source, and the `carthage-support` branch).

I suspect it has something to do with `xcuserdata` of the Desktop Viewer project, take a look [here](https://github.com/victorgama/NSLogger/tree/carthage-support/Desktop%20Viewer/NSLoggerViewer.xcodeproj/xcuserdata), it contains a directory with my local username, that appeared after I opened the project or used the script I also mentioned [here](https://github.com/fpillet/NSLogger/pull/184#issuecomment-175874330), and inside this directory you will find schemes and such.

I also tried to purge those files, and move NSLogger shared schemes to the xcworkspace, but Carthage still tries to compile the Desktop Viewer nevertheless.

TL;DR: I don't know. It may need more knowledge of Xcode inner workings than I have. :(

We can also try to remove the shared schemes from the viewer project, but I don't know whether @fpillet needs them. :stuck_out_tongue:

**Comment by fpillet** [comment]
I think you can remove shared schemes or tweak them as needed - I don't have specific requirements to build the release versions of NSLogger. I'd rather see working Carthage support at this stage :)

**Comment by victorgama** [comment]
Thanks @fpillet! Will work on it right now. :)

**Comment by victorgama** [comment]
Hey @oflannabhra, mind trying to build tag `v1.7.1` from my fork? It should be built by default if none specified, so just `github "victorgama/nslogger"` should suffice. :shipit:

**Comment by oflannabhra** [comment]
@victorgama :tada: Success! :tada: 
```
± carthage update
*** Fetching NSLogger
*** Checking out NSLogger at "v1.7.1"
*** xcodebuild output can be found in /var/folders/x7/h2hsk83d7x1_7cwwmqvx1klm0000gn/T/carthage-xcodebuild.BiGqUA.log
*** Building scheme "NSLogger" in NSLogger.xcodeproj
** BUILD SUCCEEDED **
```

**Comment by victorgama** [comment]
@oflannabhra Great! I just rebased my changes from the `test` branch (to which those tags were pointing to) to the `carthage-support` branch. I think we're ready to merge, @fpillet! :shipit: :confetti_ball:

**Comment by oflannabhra** [comment]
:+1: for merge!

**Comment by fpillet** [comment]
Thanks guy, great work! :+1:

**Comment by victorgama** [comment]
Thank you, @fpillet!
Also kudos to @oflannabhra, for helping build it! 🤘🏻

**Comment by oflannabhra** [comment]
@victorgama, maybe I'm missing something here, but after the PR merge I can't use Carthage to include NSLogger.

Carthage's build of `NSLogger.xcodeproj` times out due to no shared schemes. I checked out `master` and opening `NSLogger.xcodeproj` in Xcode error out due to a lack of `project.pbxproj` file. Did something get lost when you rebased? Or was something gitignored?

**Comment by victorgama** [comment]
@oflannabhra It makes sense, the last tag @fpillet pushed is from Dec. 2, 2015. This tag does not contain the changes I made, and is picked by Carthage to be built.<sup>[[1]](https://github.com/Carthage/Carthage#tag-stable-releases)</sup>

Until a new tag is pushed, one workaround is to use the following line on your Cartfile:

```
github "fpillet/nslogger" "master"
```

Or, if you don't feel comfortable using the last commit made on master, point directly to the commit that merges this PR:

```
github "fpillet/NSLogger" "41f2e08"
```

Hope this helps! :smiley:

**Comment by fpillet** [comment]
Ha interesting! So if I push a new tag, you guys won't have to pull your hair to get Carthage to work with NSLogger? Neat. Will do today!

**Comment by victorgama** [comment]
@fpillet Pointing to a branch/commit also works, but tagging is the way to go. 😀

Thank you!

**Comment by victorgama** [comment]
@fpillet Friendly reminder to push a new tag.
Until you push it, the documentation regarding Carthage compatibility is quite misleading.

**Comment by fpillet** [comment]
Done! tagged 1.6.1, should be good for you. Sorry for the delay!

**Comment by victorgama** [comment]
Thank you!
